File: lab7/lab7/repository/InMemoryRepository.cs
using lab7.domain;
using lab7.domain.validator;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace lab7.repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {
        protected IValidator<E> validator;

        protected IDictionary<ID, E> entities = new Dictionary<ID, E>();

        public InMemoryRepository(IValidator<E> validator)
        {
            this.validator = validator;
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values.ToList<E>();
        }

        public E FindOne(ID id)
        {
            return entities[id];
        }

        public E Save(E entity)
        {
            if (entity == null)
                throw new ArgumentNullException("entity must not be null");
            this.validator.Validate(entity);
            if (this.entities.ContainsKey(entity.ID))
            {
                return entity;
            }
            this.entities[entity.ID] = entity;
            return default(E);
        }
    }
}

File: vacante/src/main/java/repository/memory/InMemoryRepository.java
package repository.memory;

import repository.Repository;
import vacante.domain.Entity;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, E extends Entity<ID>> implements Repository<ID, E> {
    Map<ID, E> entities;

    public InMemoryRepository() {
        entities = new HashMap<ID, E>();
    }

    @Override
    public E findOne(ID id) {
        if(id == null)
            throw new IllegalArgumentException("id must not be null");
        return entities.get(id);
    }

    @Override
    public Iterable<E> findAll() {
        return entities.values();
    }

    @Override
    public E add(E entity) {
        if(entity == null)
            throw new IllegalArgumentException("entity must not be null");
        if(entities.get(entity.getId()) != null)
            return entity;
        else entities.put(entity.getId(), entity);
        return null;
    }
}
